### ISSUE TITLE
Cannot instantiate meta to solution...: Pattern checking done too early in where block

### ISSUE THREAD

**Issue by andreasabel** [agda/agda#2056]
``` agda
data Size : Set where

↑ : Size → Size
↑ ()

data N : Size → Set where
  suc : ∀{i} (a : N i) → N (↑ i)

data Val : ∀{i} (t : N i) → Set where
  val : ∀{i} (n : N i) → Val (suc n)

record R (j : Size) : Set where
  field num : N j

data W {j} (ft : R j) : Set where
  immed : (v : Val (R.num ft)) → W ft

postulate
  E : ∀{j} (ft : R j) (P : (w : W ft) → Set) → Set

test : ∀ {j} (ft : R j) → Set
test {j} ft = E {j} ft testw
  where
  testw : ∀ {ft : R _} (w : W ft) → Set
  testw (immed (val a)) = test record{ num = a }

-- testw passes without quantification over ft
-- or with _ := j
```

reports

```
Cannot instantiate the metavariable _35 to solution ↑ i since it
contains the variable i which is not in scope of the metavariable
or irrelevant in the metavariable but relevant in the solution
when checking that the pattern val a has type Val (R.num ft₁)
```

The problem here is that the unifier cannot deal with the unsolved meta.

The meta can be solved by the call to `testw` outside the `where` block.
Maybe the checking of the `where` block should only done that far until the parent rhs can be checked.  Only then, the remaining of the `where` block would be checked.

In the end, this is an instance of `Postponing checking of left hand sides` (see #480).

But maybe, there is also some bug that causes the instantiation to fail?!

**Comment by andreasabel** [comment]
@jespercockx Can you take a look?

P.S.: Agda < 2.5 gives error "Refuse to solve heterogeneous constraint".
Even if the meta cannot be solved here, there should be a better error message.

**Comment by jespercockx** [comment]
When checking just the where block (before the meta is solved by the call to testw in the main clause), Agda encounters the unification problem `↑ i =?= _35 ft`. In theory, this equation can be solved if _35 is either a flexible variable (e.g. `j`) or if it is equal to `↑ i`.  But consider that there could be multiple flexible variables of type `Size` in scope, and all of them could equally well lead to a sensible solution. So there is no way to guess that the meta should be `j` by only considering the where block. So the unifier tries the only other possibility: instantiating the meta to `↑ i`, which fails since `i` is bound locally in the clause and hence not in scope of the metavariable. So in conclusion, I don't think this can be solved without first solving #480.

**Comment by jespercockx** [comment]
I don't understand the current status of #480. It says it's fixed, but I cannot find any machinery for postponing lhs checking. Shouldn't there be some constructor for `Constraint` or `PostponedTypeCheckingProblem` for a postponed LHS?

**Comment by andreasabel** [comment]
Only the original issue of #480 is fixed.  In general, there is no postponing of LHS checking.

**Comment by andreasabel** [comment]
Would it not make sense to apply the substitution also to the current meta-variable arguments?  So if meta `_35` is dependent on a variable that is refined by the unifier via a substitution, should not the dependency of `_35` be updated accordingly?

**Comment by jespercockx** [comment]
I don't understand what you mean, the substitution _is_ applied to the meta-variable arguments. The problem's just that Agda cannot find a substitution because it cannot guess solutions out of nowhere (or out of the parent clause of a where block, in this case).

**Comment by jespercockx** [comment]
On the subject of improving the error message: I think we would get better error messages if we didn't instantiate metavariables so eagerly during unification. For example, if I change line 757 of Unify.hs from
`addContext (varTel s) $ noConstraints $ equalTerm a u v`
to
`addContext (varTel s) $ noConstraints $ dontAssignMetas $ equalTerm a u v`
then the error message becomes

```
Failed to solve the following constraints:
  [44] ↑ i = _35 ft : Size
when checking that the pattern val a has type Val (R.num ft₁)
```

However, this change breaks a few other testcases (issue #199, issue #314, and issue #501). I think it may be worth it to make this change anyway, since the solutions found in these examples are non-unique anyway.

**Comment by andreasabel** [comment]
I agree for #314, but why is the solution T := D in #199 not unique?

**Comment by jespercockx** [comment]
Ok, I agree the solution T := D in #199 is unique. But in general the mechanism that instantiates metavariables during unification doesn't ensure this, so I'm not sure if we should keep it.

**Comment by mietek** [comment]
```
Cannot instantiate the metavariable _252 to solution xs since it
contains the variable xs which is not in scope of the metavariable
or irrelevant in the metavariable but relevant in the solution
when checking that the expression g y ? ? has type y ∈ xs
```

I hope we can also agree this error message needs to be taken out and shot.  It would be equally useful if it said “Something went wrong. Try rearranging your implicits.”

**Comment by jespercockx** [comment]
Yes, I agree that's one of Agda's least helpful error messages (and that's saying something). The problem is that a better error message would depend on a lot of non-local information that we just don't have available right now:
- The reason why the metavariable _252 was introduced (an implicit argument, an underscore, a postponed constraint, ...).
- In case the meta was introduced because of an implicit argument, the source location of the (hopefully visible?) function.
- Whether the argument `xs` of the metavariable was pruned or it was never in scope of the metavariable in the first place.
- In the former case, the reason why it was pruned.

Unless we keep (at least some of) this information, we're stuck with error messages no more helpful than "Something went wrong".

**Comment by nad** [comment]
@jespercockx wrote:

> in general the mechanism that instantiates metavariables during unification doesn't ensure [that solutions are unique]

This sounds like a bug to me.

**Comment by jespercockx** [comment]
I think so too, however it doesn't impact soundness and fixing it could break people's code, so that's why I've kept the old behaviour wrt metavariables in the new unifier so far. Maybe I can push a fix to master instead of stable so people have some time to adapt before the change is released?

**Comment by jespercockx** [comment]
I think we are discussing three different topics here:
1. We should be able to postpone LHS checking in order to check the example in the OP.
2. We should give better error messages when a metavariable cannot be instantiated.
3. The unifier shouldn't instantiate metavariables since this can lead to non-unique solutions.

I propose we continue the discussion of 1. here, I'll create new issues for 2. and 3.

**Comment by jespercockx** [comment]
New error message after the fix of #2074:

```
Failed to solve the following constraints:
  [44] ↑ i = _j_39 ft : Size
when checking that the pattern val a has type Val (R.num ft₁)
```
